### PR TITLE
Skeletal input

### PR DESCRIPTION
**Pull request by artumino**
Current Issues:

- [x] Controllers are not aligned with the hands
- [x] Tools and Pickup are not aligned with the hands (will fix with poses)
- [ ] Hand motion range doesn't always update (probably solved in latest commit)
- [x] Controllers seems to persist in hands after one timeloop ends

### CODE REVIEW

**Comment by Raicuparta** [comment]
> Controllers seems to persiste in hands after one timeloop ends

@artumino Actually this bug already existed before. Happens because I'm not triggering the visibility change in all the correct times. I just never bothered to fix it. I can help fixing it in a separate PR if you want.

**Comment by artumino** [comment]
@Raicuparta No problem then, it was annoying cause the RangeOfMotion for the skeleton also wasn't working and the hand could clip through the controllers. It should work now so no more visual glitches even when the controllers are enabled (i think).

The hand-controller missalignment is probably due to the retargetting so we have to live with it (or maybe try getting the hand working directly without the retargeting).

I experimented with poses on another branch and aligning tools is just a matter of repositioning each Holdable object again. I've added a transform in the middle of the palm so calculating offsets should be easy.

**Comment by Raicuparta** [comment]
@artumino I'm not sure what you mean by retargeting. I just tested, and the controllers seem to be aligned to the hands? Or maybe I just don't know what I'm looking for.

I also tested holding the tools and objects, and yeah it clips a bit more to the fingers now but that's not a big deal, can definitely include it as-is in an update.

I noticed that the wrist "hole" in the models is very shiny, probably because of the specular map texture or something similar. It makes that part kinda glow in the dark. Would be better to make it completely matte.

File: NomaiVR/Input/VRActionInput.cs [hunk 1]
```diff
@@ -141,13 +141,13 @@ private bool IsOppositeHandWithSameName(VRActionInput other)
...
-        private string[] GetHoldActionText()
+        private List<string>GetHoldActionText()
         {
             if (_holdActionInput == null)
             {
-                return new string[] { };
+                return new List<string>();
             }
-            return _holdActionInput.GetText();
+            return _holdActionInput.GetText().ToList();
```
Comment (Raicuparta) [inline NomaiVR/Input/VRActionInput.cs:150]: Why this change?

File: NomaiVR/UI/Dialogue.cs [hunk 2]
```diff
@@ -11,21 +11,23 @@ internal class Dialogue : NomaiVRModule<Dialogue.Behaviour, Dialogue.Behaviour.P
...
+                _uiGameObject = GameObject.FindWithTag("DialogueGui");
+                _canvasTransform = _uiGameObject.transform;//GameObject.Find("DialogueCanvas").transform;
 
                 _canvasTransform.localScale *= _dialogeRenderSize;
 
                 // Prevent dialogue box from flying off after a while.
-                _canvasTransform.parent = new GameObject().transform;
-                _canvasTransform.parent.gameObject.AddComponent<FollowTarget>().target = Locator.GetPlayerTransform();
+                _canvasTransform.parent = Locator.GetPlayerTransform();
+                //_canvasTransform.parent.gameObject.AddComponent<FollowTarget>().target = Locator.GetPlayerTransform();
```
Comment (Raicuparta) [inline NomaiVR/UI/Dialogue.cs:28]: I think there was some reason why I had to use a FollowTarget here instead of just normal parenting? It was breaking something. Should have left a comment explaining it. I'll investigate a bit to see why that was there.

Comment (artumino) [reply NomaiVR/UI/Dialogue.cs:28]: Yep these changes come from a test I was making I forgot to clean them up for the PR.
It seems to work fine for now, I'll revert them for this PR since they are out of scope.

Comment (Raicuparta) [reply NomaiVR/UI/Dialogue.cs:28]: It does actually seem to work fine and is obviously much cleaner. But I agree, better leave as it was to avoid breaking something 🤷‍♂️

File: NomaiVR/Input/VRActionInput.cs (hunk 1 above)
Comment (artumino) [reply NomaiVR/Input/VRActionInput.cs:150]: This is cause I updated Harmony and the Add to a string array at line 56 was an Extension method in Harmony.
The newer version changed namespace and name of the method to AddToArray. At the time I didn't look into Harmony's documentation. 

I can update it to use the new extension method but I notice that this is the only place where we use Harmony directly, everything else is done through OWML. I also looked at the code for Hamony's extension and it's doing effectively the same thing in memory. It's up to you to decide, by keeping this change Harmony can be removed as a dependecy.

I could also leave GetHoldActionText unchanged and replace Add with Harmony's implementation:
`return GetHoldActionText().Concat(new[] { WrapWithBrackets(result) }).ToArray();`

Comment (Raicuparta) [reply NomaiVR/Input/VRActionInput.cs:150]: Ah ok, understood. The harmony dependency thing is a leftover from when OWML didn't have some features it now has. I see how it doesn't make sense anymore. I'll leave it up to you.

**Comment by artumino** [comment]
> 
> 
> @artumino I'm not sure what you mean by retargeting. I just tested, and the controllers seem to be aligned to the hands? Or maybe I just don't know what I'm looking for.
> 
> I also tested holding the tools and objects, and yeah it clips a bit more to the fingers now but that's not a big deal, can definitely include it as-is in an update.
> 
> I noticed that the wrist "hole" in the models is very shiny, probably because of the specular map texture or something similar. It makes that part kinda glow in the dark. Would be better to make it completely matte.

Yeah I used the wrong term, the controllers are aligned. What isn't perfectly aligned is the fingertips with respect to the controller buttons, for example in the main menu if I move the thumbstick the thumb doesn't align with the thumbstick on the rendered controller. That's because we are retargeting the solved skeleton from SteamVR to a bigger skeleton for the hearthian hand, which means that we simply copying the rotation for each bone in the source human hand skeleton to the slightly bigger hearthian one and we endup with fingertips in a different location.

As I said I think we can live with this mismatch, it's only really visible in the main menu. To completely fix the issue I think we would have to make a 5 fingered skeleton for the hand with the heartian proportions, ignore 2 fingers during skinning and hope that steamvr can solve finger position with it, that would give us the right rotations to apply without any retargetting involved.

As for the glove, yes that part is currently mapped to the black rubber used at the tip of the fingers, I should probably make a new set of textures for this gloves, since right now I'm borrowing the one from the suit.

**Comment by Raicuparta** [comment]
@artumino ah yes I think we don't need to worry about the fingers aligning perfectly. It's an alien and all, people will understand :P

The shiny wrist hole, i was talking about the non-glove hands. I didn't notice any material issues with the gloves.

**Comment by artumino** [comment]
> 
> 
> @artumino ah yes I think we don't need to worry about the fingers aligning perfectly. It's an alien and all, people will understand :P
> 
> The shiny wrist hole, i was talking about the non-glove hands. I didn't notice any material issues with the gloves.

Oh, ok. If there are no issues with the gloves then I'll look at the specular for the hand. That surely is easier to edit :P

File: NomaiVR/manifest.json [hunk 3]
```diff
@@ -4,7 +4,7 @@
   "name": "NomaiVR",
   "uniqueName": "Raicuparta.NomaiVR",
   "description": "Outer Wilds VR mod, with full motion control support.",
-  "version": "1.0.2",
-  "owmlVersion": "1.0.1",
+  "version": "1.0.3",
```
Comment (Raicuparta) [inline NomaiVR/manifest.json:7]: Please revert this change, NomaiVR's version will be bumped in a dedicated commit. Also, this is a new feature, so we'll bump the minor instead of patch version (1.1.0).

File: NomaiVR/Assets/AssetLoader.cs [hunk 4]
```diff
@@ -25,9 +30,10 @@ public AssetLoader()
             PostCreditsPrefab = LoadAsset<GameObject>(postCreditsBundle, "postcreditscamera.prefab");
             PostCreditsRenderTexture = LoadAsset<RenderTexture>(postCreditsBundle, "screen.renderTexture");
 
-            var handsBundle = LoadBundle("hands");
-            HandPrefab = LoadAsset<GameObject>(handsBundle, "righthandprefab.prefab");
-            GlovePrefab = LoadAsset<GameObject>(handsBundle, "rightgloveprefab.prefab");
+            //var handsBundle = LoadBundle("hands");
+            var skeletalHandsBundle = LoadBundle("skeletal-hands");
+            HandPrefab = LoadAsset<GameObject>(skeletalHandsBundle, "Assets/skeletal_hand.prefab");
+            //GlovePrefab = LoadAsset<GameObject>(skeletalHandsBundle, "Assets/skeletal_glove.prefab");
```
Comment (Raicuparta) [inline NomaiVR/Assets/AssetLoader.cs:36]: cleanup commented lines?

File: NomaiVR/Hands/Hand.cs [hunk 5]
```diff
@@ -1,58 +1,218 @@
...
+                action_bool = SteamVR_Actions.default_GrabPinch,
+                enabled = true,
+                influence = 1,
+                name = "fist",
+                pose = 2,
+                value = 0,
+                type = SteamVR_Skeleton_Poser.PoseBlendingBehaviour.BlenderTypes.BooleanAction,
+                previewEnabled = true,
+                smoothingSpeed = 16
+            });
```
Comment (Raicuparta) [inline NomaiVR/Hands/Hand.cs:195]: My interpretation of this is that, if there's no skeleton available, it will use the specified buttons to define these poses. I tested removing the skeleton definitions from my personal SteamVR input bindings, but pressing Grip didn't seem to trigger the "point" gesture as I expected. Am I misinterpreting this? The hands looked fine with the skeleton disabled though, so it's absolutely fine if that's what's supposed to happen too.

File: NomaiVR/Hands/HandsController.cs [hunk 6]
```diff
@@ -74,21 +76,27 @@ private void SetUpHands()
...
                 RightHand = right.transform;
+                RightHandBehaviour = right;
 
                 var left = new GameObject().AddComponent<Hand>();
                 left.pose = SteamVR_Actions.default_LeftHand;
                 left.transform.parent = _wrapper;
-                left.transform.localPosition = new Vector3(-0.03f, 0.05f, -0.2f);
-                left.transform.localRotation = Quaternion.Euler(313f, 350f, 65f);
+                left.transform.localPosition = Vector3.zero;//new Vector3(-0.03f, 0.05f, -0.2f);
+                left.transform.localRotation = Quaternion.identity; //Quaternion.Euler(313f, 350f, 65f);
```
Comment (Raicuparta) [inline NomaiVR/Hands/HandsController.cs:92]: remove commented out code?

File: NomaiVR/Input/VRActionInput.cs [hunk 7]
```diff
@@ -52,8 +51,7 @@ public string[] GetText()
             {
                 return new string[] { };
             }
-            var holdInputText = GetHoldActionText();
-            return holdInputText.Add(WrapWithBrackets(result)).ToArray();
+            return GetHoldActionText().Concat(new[] { WrapWithBrackets(result) }).ToArray();
```
Comment (Raicuparta) [inline NomaiVR/Input/VRActionInput.cs:54]: I think you said you would revert these since they're out of scope? But if you're sure the result is the same, I'm fine with it. I'm just always nervous about the input stuff since it has always been super fragile :P

**Review by Raicuparta — APPROVED** [review]
